internal/backend/webhookssvc/svc.go

```diff
@@ -100,77 +105,73 @@ func requestToData(r *http.Request) (map[string]sdktypes.Value, error) {
+		"method":  sdktypes.NewStringValue(r.Method),
+		"raw_url": sdktypes.NewStringValue(r.URL.String()),
+		"url":     urlData(r.URL),
 	}, nil
 }
 
-func bodyToStruct(body []byte, form url.Values) sdktypes.Value {
-	var (
-		v        any
-		jsonBody sdktypes.Value
+func bodyData(body []byte, form url.Values) sdktypes.Value {
+	bytes := sdktypes.Nothing
+	if len(body) > 0 {
+		bytes = sdktypes.NewBytesValue(body)
+	}
+
+	return sdktypes.NewDictValueFromStringMap(
+		map[string]sdktypes.Value{
+			"bytes": bytes,
+			"form":  formData(form),
+			"json":  jsonData(body),
+		},
+	)
+}
+
+func formData(form url.Values) sdktypes.Value {
+	if len(form) == 0 {
+		return sdktypes.Nothing
+	}
+
+	return sdktypes.NewDictValueFromStringMap(
+		kittehs.TransformMapValues(form, func(vs []string) sdktypes.Value {
+			return sdktypes.NewStringValue(strings.Join(vs, ", "))
+		}),
 	)
+}
 
-	jsonDecoder := json.NewDecoder(bytes.NewReader(body))
-	jsonDecoder.UseNumber()
+func jsonData(body []byte) sdktypes.Value {
+	var v any
+	d := json.NewDecoder(bytes.NewReader(body))
+	d.UseNumber()
 
-	if err := jsonDecoder.Decode(&v); err != nil {
-		jsonBody = kittehs.Must1(sdktypes.NewConstFunctionError("json", err))
+	if err := d.Decode(&v); err != nil {
+		return sdktypes.Nothing
 	} else if vv, err := sdktypes.WrapValue(v); err != nil {
-		jsonBody = kittehs.Must1(sdktypes.NewConstFunctionError("json", err))
+		return sdktypes.Nothing
 	} else {
-		jsonBody = kittehs.Must1(sdktypes.NewConstFunctionValue("json", vv))
+		return vv
 	}
+}
 
-	// add form() only for requests (when not nil) and not for responses
-	if form != nil {
-		formBody := kittehs.Must1(sdktypes.NewConstFunctionValue("form", sdktypes.NewDictValueFromStringMap(
-			kittehs.TransformMapValues(form, func(vs []string) sdktypes.Value {
-				return sdktypes.NewStringValue(strings.Join(vs, ","))
-			}),
-		)))
-		return kittehs.Must1(sdktypes.NewStructValue(
-			sdktypes.NewStringValue("body"),
-			map[string]sdktypes.Value{
-				"text":  kittehs.Must1(sdktypes.NewConstFunctionValue("text", sdktypes.NewStringValue(string(body)))),
-				"bytes": kittehs.Must1(sdktypes.NewConstFunctionValue("bytes", sdktypes.NewBytesValue(body))),
-				"json":  jsonBody,
-				"form":  formBody,
-			},
-		))
-	}
-	return kittehs.Must1(sdktypes.NewStructValue(
-		sdktypes.NewStringValue("body"),
+func urlData(u *url.URL) sdktypes.Value {
+	return sdktypes.NewDictValueFromStringMap(
 		map[string]sdktypes.Value{
-			"text":  kittehs.Must1(sdktypes.NewConstFunctionValue("text", sdktypes.NewStringValue(string(body)))),
-			"bytes": kittehs.Must1(sdktypes.NewConstFunctionValue("bytes", sdktypes.NewBytesValue(body))),
-			"json":  jsonBody,
+			"fragment": sdktypes.NewStringValue(u.Fragment),
+			"path":     sdktypes.NewStringValue(u.Path),
+			"query": sdktypes.NewDictValueFromStringMap(
+				kittehs.TransformMapValues(u.Query(), func(vs []string) sdktypes.Value {
+					return sdktypes.NewStringValue(strings.Join(vs, ", "))
```

Comment:
```suggestion
					return sdktypes.NewStringValue(strings.Join(vs, ","))
```

Reply:
afacis, canonical implementations and examples of HTTP don't distinguish between:
```
Header: foo, bar
```
And:
```
Header: foo
Header: bar
```
Example: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding